sql: fix inverted index queries

justinj opened the pull request:
We need to do a full table scan whenever we are querying by containment
of a document that has an empty object or array in it, not just those at
the top level.

Fixes #23819.

Release note (bug fix): Fixed a bug where some inverted index queries
could return incorrect results.

Commit:
sql: fix inverted index queries

We need to do a full table scan whenever we are querying by containment
of a document that has an empty object or array in it, not just those at
the top level.

Fixes #23819.

Release note (bug fix): Fixed a bug where some inverted index queries
could return incorrect results.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/23930)
<!-- Reviewable:end -->

jordanlewis commented:
Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[pkg/sql/logictest/testdata/logic_test/inverted_index, line 488 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23930#-L7fjPimAI7SUy91BS3D:-L7fjPimAI7SUy91BS3E:bf525l2) ([raw file](https://github.com/cockroachdb/cockroach/blob/07813aff914f204b2614997679166e5dbc9b47a0/pkg/sql/logictest/testdata/logic_test/inverted_index#L488)):*
> ```
> 
> query IT
> SELECT * from d where b @> '{"a": {}}' ORDER BY a;
> ```

Missing a test case for empty object contained within array.

---

*[pkg/sql/opt/idxconstraint/index_constraints.go, line 868 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23930#-L7fiM7dBVF6wB6eFydj:-L7fiM7dBVF6wB6eFydk:b-lynrhq) ([raw file](https://github.com/cockroachdb/cockroach/blob/07813aff914f204b2614997679166e5dbc9b47a0/pkg/sql/opt/idxconstraint/index_constraints.go#L868)):*
> ```Go
> 				// Errors here can only occur on invalid encoded data, but the rhs here
> 				// should always be a constant.
> 				panic(fmt.Sprintf("programming error: %s", err))
> ```

how sure are you of this? could also just return `nil, false, false` and log an error.

---

*[pkg/util/json/encoded.go, line 705 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23930#-L7fiTFu2ML368zlnpwd:-L7fiTFu2ML368zlnpwe:b-w879q9) ([raw file](https://github.com/cockroachdb/cockroach/blob/07813aff914f204b2614997679166e5dbc9b47a0/pkg/util/json/encoded.go#L705)):*
> ```Go
> }
> 
> // EncodeInvertedIndexKeys implements the JSON interface.
> ```

wrong comment title

---

*[pkg/util/json/json.go, line 1535 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23930#-L7fipSo2ooN94aTR6R2:-L7fipSo2ooN94aTR6R3:b-91lkrg) ([raw file](https://github.com/cockroachdb/cockroach/blob/07813aff914f204b2614997679166e5dbc9b47a0/pkg/util/json/json.go#L1535)):*
> ```Go
> 	}
> 	for _, c := range j {
> 		next, err := c.v.HasContainerLeaf()
> ```

next is a weird thing to call this, isn't it?

---

*[pkg/util/json/json_test.go, line 1644 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23930#-L7fj4m09OhiKZ_RUK2m:-L7fj4m09OhiKZ_RUK2n:b-lr66e7) ([raw file](https://github.com/cockroachdb/cockroach/blob/07813aff914f204b2614997679166e5dbc9b47a0/pkg/util/json/json_test.go#L1644)):*
> ```Go
> 		{`{}`, true},
> 		{`{"a": 1}`, false},
> 		{`{"a": {}}`, true},
> ```

This is missing test cases with empty container leaves in the middle of the tree. NBD but might as well just double check that there's no dumb regression there later. 

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23930)*
<!-- Sent from Reviewable.io -->

justinj commented:
Review status: 1 of 6 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/sql/logictest/testdata/logic_test/inverted_index, line 488 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23930#-L7fjPimAI7SUy91BS3D:-L7fm668V8HBlUEgwrEl:blsmmqo) ([raw file](https://github.com/cockroachdb/cockroach/blob/07813aff914f204b2614997679166e5dbc9b47a0/pkg/sql/logictest/testdata/logic_test/inverted_index#L488)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Missing a test case for empty object contained within array.
</blockquote></details>

there's one on line 420, did you mean empty array within object?  added.

---

*[pkg/sql/opt/idxconstraint/index_constraints.go, line 868 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23930#-L7fiM7dBVF6wB6eFydj:-L7fn7ot6Vfcwj-vgAuL:b-njoqls) ([raw file](https://github.com/cockroachdb/cockroach/blob/07813aff914f204b2614997679166e5dbc9b47a0/pkg/sql/opt/idxconstraint/index_constraints.go#L868)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

how sure are you of this? could also just return `nil, false, false` and log an error.
</blockquote></details>

I'm completely sure that it can only happen on invalid encoded data, less sure the RHS is always a constant. Do we ever do that? Just log an error but carry on? I don't particularly mind, though.

---

*[pkg/util/json/encoded.go, line 705 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23930#-L7fiTFu2ML368zlnpwd:-L7fnx7dVrONDL_T3oK6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/07813aff914f204b2614997679166e5dbc9b47a0/pkg/util/json/encoded.go#L705)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

wrong comment title
</blockquote></details>

Done.

---

*[pkg/util/json/json.go, line 1535 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23930#-L7fipSo2ooN94aTR6R2:-L7fnxluxLbAvY_7TiPV:bwfglp) ([raw file](https://github.com/cockroachdb/cockroach/blob/07813aff914f204b2614997679166e5dbc9b47a0/pkg/util/json/json.go#L1535)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

next is a weird thing to call this, isn't it?
</blockquote></details>

I guess so, `child` might be more appropriate.

---

*[pkg/util/json/json_test.go, line 1644 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23930#-L7fj4m09OhiKZ_RUK2m:-L7fo9e1IUh6Ntn1sfa1:b1eh31l) ([raw file](https://github.com/cockroachdb/cockroach/blob/07813aff914f204b2614997679166e5dbc9b47a0/pkg/util/json/json_test.go#L1644)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This is missing test cases with empty container leaves in the middle of the tree. NBD but might as well just double check that there's no dumb regression there later. 
</blockquote></details>

I don't fully know what you mean by the "middle", but I added some more test cases with >1-ary nodes

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23930)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 1 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/logictest/testdata/logic_test/inverted_index, line 514 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23930#-L7j8p5mJK7y0Pl5pzo-:-L7j8p5mJK7y0Pl5pzo0:bcme6wq) ([raw file](https://github.com/cockroachdb/cockroach/blob/16d50f73f1aedd9ede532fc8a9316b01e2cde142/pkg/sql/logictest/testdata/logic_test/inverted_index#L514)):*
> ```
> 
> statement ok
> CREATE TABLE users (
> ```

One more thing - please add an explain test or two that demonstrates that these get turned into full table scans.

---

*[pkg/sql/opt/idxconstraint/index_constraints.go, line 868 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23930#-L7fiM7dBVF6wB6eFydj:-L7j91fz3pnVx5aqNNsr:b6juy2h) ([raw file](https://github.com/cockroachdb/cockroach/blob/07813aff914f204b2614997679166e5dbc9b47a0/pkg/sql/opt/idxconstraint/index_constraints.go#L868)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

I'm completely sure that it can only happen on invalid encoded data, less sure the RHS is always a constant. Do we ever do that? Just log an error but carry on? I don't particularly mind, though.
</blockquote></details>

@raduberinde can the RHS here ever be a non-constant?

---

*[pkg/util/json/json_test.go, line 1644 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23930#-L7fj4m09OhiKZ_RUK2m:-L7j98JtX5uhaFtbYOV5:bt69qy2) ([raw file](https://github.com/cockroachdb/cockroach/blob/07813aff914f204b2614997679166e5dbc9b47a0/pkg/util/json/json_test.go#L1644)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

I don't fully know what you mean by the "middle", but I added some more test cases with >1-ary nodes
</blockquote></details>

I meant leaves that aren't at the far "right" of the tree. You satisifed my ask. Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23930#-:-L7j9HFrs0hg2U8olrio:bnfp4nl)*
<!-- Sent from Reviewable.io -->
